pyteal/ast/opup.py

```diff
@@ -87,35 +97,49 @@ def __init__(self, mode: OpUpMode, target_app_id: Expr = None):
 
         self.mode = mode
 
-    def _construct_itxn(self) -> Expr:
+    def _construct_itxn(self, inner_fee: Expr = None) -> Expr:
+        fields: dict[TxnField, Expr | list[Expr]] = {
+            TxnField.type_enum: TxnType.ApplicationCall
+        }
+
+        # If an inner_fee is specified
+        # add it to the transaction fields
+        if inner_fee is not None:
+            require_type(inner_fee, TealType.uint64)
+            fields[TxnField.fee] = inner_fee
+
         if self.mode == OpUpMode.Explicit:
-            return Seq(
-                InnerTxnBuilder.Begin(),
-                InnerTxnBuilder.SetFields(
-                    {
-                        TxnField.type_enum: TxnType.ApplicationCall,
-                        TxnField.application_id: self.target_app_id,
-                    }
-                ),
-                InnerTxnBuilder.Submit(),
-            )
+            fields |= {TxnField.application_id: self.target_app_id}
         else:
-            return Seq(
-                InnerTxnBuilder.Begin(),
-                InnerTxnBuilder.SetFields(
-                    {
-                        TxnField.type_enum: TxnType.ApplicationCall,
-                        TxnField.on_completion: OnComplete.DeleteApplication,
-                        TxnField.approval_program: ON_CALL_APP,
-                        TxnField.clear_state_program: ON_CALL_APP,
-                    }
-                ),
-                InnerTxnBuilder.Submit(),
-            )
+            fields |= {
+                TxnField.on_completion: OnComplete.DeleteApplication,
+                TxnField.approval_program: ON_CALL_APP,
+                TxnField.clear_state_program: ON_CALL_APP,
+            }
 
-    def ensure_budget(self, required_budget: Expr) -> Expr:
+        return InnerTxnBuilder.Execute(fields)
+
+    def ensure_budget(self, required_budget: Expr, inner_fee: Expr = None) -> Expr:
```

Comment:
@barnjamin + reviewers:  In the prior PR, @cusma mentioned changing the default behavior for security benefits.  How (if at all) should the PR reflect the security concern?

* The concern is:  The default behavior (no fee specified) charges fees to the app's account when the caller does not overpay.  Such behavior may _unintentionally_ drain the app's account.
* Although the PR is a draft, I raise the note now because the outcome _may_ change PR direction.  There's at least a couple of options:
  * Make no behavior change, and update the method docs to clearly highlight the risk.
  * Agree the concern warrants different behavior.  Deprecate existing methods and offer new methods with a safer default _or_ require all parameters (avoid defaults).
* My 2c:  
  * If we started fresh, I think requiring all parameters is most robust.
  * Since we have an existing API, I _think_ doc changes are sufficient.  Though I'd like for folks more attuned to the problem frequency to chime in.

Reply:
Right, lets give the tools to make the desired behavior easier and improve docs to make it more clear

Reply:
Once the code seems right, I'll work on docs

Reply:
@barnjamin Resolving - Thanks for the reply.  I trust your judgment that extending the docs is sufficient.